src/lib/userSponsorAuthCall/index.ts

```diff
@@ -0,0 +1,160 @@
+import axios from "axios";
+import { ethers } from "ethers";
+
+import { GELATO_RELAY_URL } from "../../constants";
+import { getHttpErrorMessage } from "../../utils";
+import {
+  PaymentType,
+  RelayRequestOptions,
+  RelayResponse,
+  SignerProfile,
+} from "../types";
+
+import {
+  generateUserSponsorSignatureWith1BalanceAndSponsor,
+  generateUserSponsorSignatureWith1BalanceAndUser,
+} from "./1balance";
+import {
+  UserSponsorAuthCallWith1BalanceRequest,
+  UserSponsorAuthCallWith1BalanceStruct,
+} from "./1balance/types";
+import {
+  generateUserSponsorSignatureWithTransferFromAndSponsor,
+  generateUserSponsorSignatureWithTransferFromAndUser,
+} from "./transferFrom";
+import {
+  UserSponsorAuthCallWithTransferFromRequest,
+  UserSponsorAuthCallWithTransferFromStruct,
+} from "./transferFrom/types";
+import {
+  SignatureResponse,
+  Signer,
+  UserSponsorAuthCallStruct,
+  UserSponsorSignatureRequest,
+} from "./types";
+
+/**
+ * @function
+ * @template PT
+ * @extends {PaymentType}
+ * @template S
+ * @extends {SignerProfile}
+ * @param {PT} paymentType - PaymentType.OneBalance or PaymentType.TransferFrom
+ * @param {S} signerProfile - SignerProfile.User or SignerProfile.Sponsor
+ * @param {UserSponsorSignatureRequest<PT, S>} request - Depending on the paymentType and signerProfile, the request to be signed by the signer
+ * @param {Signer<S>} signer - Depending on the signerProfile, Wallet or Web3Provider to sign the payload
+ * @returns {Promise<SignatureResponse>} Response body with signature and signed struct
+ *
+ */
+export const generateUserSponsorSignature = async <
+  PT extends PaymentType,
+  S extends SignerProfile
+>(
+  paymentType: PT,
+  signerProfile: S,
+  request: UserSponsorSignatureRequest<PT, S>,
+  signer: Signer<S>
+): Promise<SignatureResponse> => {
+  switch (paymentType) {
+    case PaymentType.OneBalance:
+      switch (signerProfile) {
+        case SignerProfile.User:
+          return await generateUserSponsorSignatureWith1BalanceAndUser(
+            request as UserSponsorAuthCallWith1BalanceRequest,
+            signer as ethers.providers.Web3Provider
+          );
+        case SignerProfile.Sponsor:
+          return await generateUserSponsorSignatureWith1BalanceAndSponsor(
+            request as UserSponsorAuthCallWith1BalanceStruct,
+            signer as ethers.Wallet
+          );
```

Comment:
why are you using `UserSponsorAuthCallWith1BalanceRequest` in case of user
and `UserSponsorAuthCallWith1BalanceRequest` in case of sponsor

Reply:
Struct has all the parameters as required, however Request has the optional parameters where we populate the fields on demand. I am forcing Sponsors to provide all parameters to ensure that the payloads signed by user and sponsor are the same. Sponsors will use the struct that is returned by the user signature generation

Reply:
ok got it 👍